File: 250314Shenfeng/250327multiplecsv.py
import os
import pandas as pd

# Folder containing the CSV files
folder_path = '/Users/w/Developer/250306 KYK Datamining/250314Shenfeng/datasets'

# Get a list of all CSV files in the folder
csv_files = [f for f in os.listdir(folder_path) if f.endswith('.csv')]

# Dictionary to store information about each file
file_info = {}

# Loop through each CSV file and collect information
for file in csv_files:
    file_path = os.path.join(folder_path, file)
    try:
        df = pd.read_csv(file_path)
        file_info[file] = {
            'columns': list(df.columns),
            'dtypes': df.dtypes.to_dict(),
            'shape': df.shape
        }
    except Exception as e:
        print(f"Error reading {file}: {e}")

# Compare the files
print("Comparing CSV files...\n")
for file, info in file_info.items():
    print(f"File: {file}")
    print(f"Columns: {info['columns']}")
    print(f"Data Types: {info['dtypes']}")
    print(f"Shape: {info['shape']}")
    print("-" * 50)

# Check if all files have the same format
columns_set = {tuple(info['columns']) for info in file_info.values()}
if len(columns_set) == 1:
    print("All files have the same columns.")
else:
    print("Files have different columns.")

# Optional: Check for differences in data types
dtypes_set = {tuple(info['dtypes'].items()) for info in file_info.values()}
if len(dtypes_set) == 1:
    print("All files have the same data types.")
else:
    print("Files have different data types.")
    
    
    
    """
    
    import pandas as pd

# File paths for the two CSV files
file1_path = '/Users/w/Developer/250306 KYK Datamining/250314Shenfeng/2023-open-data-dfb-ambulance.csv'
file2_path = '/Users/w/Developer/250306 KYK Datamining/250314Shenfeng/another-file.csv'  # Replace with the second file's name

try:
    # Load the two CSV files
    df1 = pd.read_csv(file1_path)
    df2 = pd.read_csv(file2_path)
    
    # Compare column names
    print("Columns in File 1:", df1.columns)
    print("Columns in File 2:", df2.columns)
    print("\nDo the files have the same columns?", list(df1.columns) == list(df2.columns))
    
    # Compare data types
    print("\nData types in File 1:")
    print(df1.dtypes)
    print("\nData types in File 2:")
    print(df2.dtypes)
    print("\nDo the files have the same data types?", df1.dtypes.equals(df2.dtypes))
    
    # Compare basic statistics
    print("\nBasic statistics for File 1:")
    print(df1.describe())
    print("\nBasic statistics for File 2:")
    print(df2.describe())
    
except Exception as e:
    print(f"An error occurred: {e}")
    """

File: 250314Shenfeng/250327multiplecsv.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <string>
#include <map>
#include <filesystem>
#include <sstream>
#include <Eigen/Dense>

namespace fs = std::filesystem;

struct CSVInfo {
    std::vector<std::string> columns;
    std::vector<std::string> dtypes;
    std::pair<int,int> shape;
};

CSVInfo analyze_csv(const std::string& file_path) {
    std::ifstream file(file_path);
    std::string line;
    CSVInfo info;
    if (std::getline(file, line)) {
        std::stringstream ss(line);
        std::string col;
        while (std::getline(ss, col, ',')) info.columns.push_back(col);
    }
    int row_count = 0;
    while (std::getline(file, line)) ++row_count;
    info.shape = {row_count, (int)info.columns.size()};
    info.dtypes = std::vector<std::string>(info.columns.size(), "string/double"); // C++ does not infer dtypes
    return info;
}

int main() {
    std::string folder_path = "250314Shenfeng/datasets";
    std::map<std::string, CSVInfo> file_info;
    for (const auto& entry : fs::directory_iterator(folder_path)) {
        if (entry.path().extension() == ".csv") {
            try {
                file_info[entry.path().filename()] = analyze_csv(entry.path().string());
            } catch (...) {
                std::cout << "Error reading " << entry.path().filename() << std::endl;
            }
        }
    }
    std::cout << "Comparing CSV files...\n";
    for (const auto& [file, info] : file_info) {
        std::cout << "File: " << file << std::endl;
        std::cout << "Columns: ";
        for (auto& c : info.columns) std::cout << c << " ";
        std::cout << std::endl;
        std::cout << "Shape: (" << info.shape.first << ", " << info.shape.second << ")" << std::endl;
        std::cout << "Dtypes: ";
        for (auto& d : info.dtypes) std::cout << d << " ";
        std::cout << std::endl << std::string(50,'-') << std::endl;
    }
    // Check if all files have the same columns
    std::set<std::vector<std::string>> columns_set;
    for (const auto& [_, info] : file_info) columns_set.insert(info.columns);
    if (columns_set.size() == 1) std::cout << "All files have the same columns." << std::endl;
    else std::cout << "Files have different columns." << std::endl;
    // Check if all files have the same dtypes
    std::set<std::vector<std::string>> dtypes_set;
    for (const auto& [_, info] : file_info) dtypes_set.insert(info.dtypes);
    if (dtypes_set.size() == 1) std::cout << "All files have the same data types." << std::endl;
    else std::cout << "Files have different data types." << std::endl;
    return 0;
}
